### DISCUSSION TITLE
Validation using submodel independently

### DISCUSSION THREAD

**Discussion by eddsalkield** [pydantic/pydantic#2671] [Question] [answered]
I'm attempting to implement a model that populates its fields dynamically by evaluating a function for each field.  The behaviour of this function may depend upon the type required by the model, and so the function should take the sub-model as an argument.

My use case for this model is to parse data of unknown type, whereby the function can try different data sources in sequence until the data matches the type, falling back to a default value if required.

A similar idea could be implemented using the ideas from [this thread](https://github.com/samuelcolvin/pydantic/issues/668), by inspecting the `ValidationError`.  However, I think that a model with a root validator should encapsulate this behaviour more cleanly.

My current idea looks like this:

```python3
from pydantic import BaseModel, root_validator

class DynamicModel(BaseModel):
    @root_validator(pre=True)
    def evaluate_functions(cls, values):
        import pdb
        pdb.set_trace()
        return {k: f(cls.__fields__[k]) for k,f in values.items()}

class Video(DynamicModel):
    url: str

video = Video(url=(lambda t: \
    'test.example.com' if t.VALIDATION_GOES_HERE('test.example.com') else 'default.example.com'))
```

From within the root validator, I can access `cls.__fields__`, returning a dict that maps submodel names to `pydantic.fields.ModelField`.  However, it doesn't seem obvious how I'd perform validation on the `ModelField` (the [`validate` method](https://github.com/samuelcolvin/pydantic/blob/62bb2ad4921016df51abf3922c3fe51113b08939/pydantic/fields.py#L691) appears to have a different, internal interface).  In my example, I've written '.VALIDATION_GOES_HERE` as a placeholder for the actual validation mechanism I should be using.

Help much appreciated :)

**Comment by eddsalkield** [comment]
After much deliberation, I have settled upon a solution, utilising a `functools.partial` to pass a reduced-interface validator to the data generation function.

I'd like to discuss whether a model with dynamically-evaluated values (like my `DynamicModel`) would be suitable for inclusion upstream, or whether this problem can be solved using a different approach that would be more suitable.

I've written an example data generation callable, `FirstSuccessfulOf`, to justify my use case.

```python3
from pydantic import BaseModel, root_validator, ValidationError
from functools import partial
from typing import _CallableGenericAlias

class DynamicModel(BaseModel):
    """
    A Pydantic model which, upon instantiation, scans each of its fields
    and evaluates any callables where their type isn't callable

    Each callable should expect two arguments:
        * a validator function, itself taking a value to validate,
          and returning a tuple (validated_value, errors)
        * cls, the class with which to construct ValidationErrors
    """
    @root_validator(pre=True)
    def evaluate_functions(cls, values):
        return {k: f(partial(cls.__fields__[k].validate, values={}, loc=k), cls)
                    # NB: This doesn't check that the callable is an
                    # exact match
                    if callable(f) and not isinstance(cls.__fields__[k].type_,
                        _CallableGenericAlias)
                else f
                for k,f in values.items()}

class FirstSuccessfulOf:
    """
    A callable which evaluates a list of values/functions, returning the first
    that successfully validates.
    If none successfully validate, returns a ValidationError containing the
    errors from each failed candidate value.
    """
    def __init__(self, *args):
        self.candidate_values = args

    def __call__(self, validate, cls):
        errors = []
        for f in self.candidate_values:
            if callable(f):
                valid_value, error = validate(f())
            else:
                valid_value, error = validate(f)
            if error:
                errors.append(error)
            else:
                return valid_value
        raise ValidationError(errors, cls)

class Video(DynamicModel):
    url: str

from typing import Callable
class C(DynamicModel):
    url: Callable[[], None]

video = Video(url=FirstSuccessfulOf(None, lambda: None, lambda: 'hello'))
video2 = Video(url='hello')

c = C(url=FirstSuccessfulOf(None, None, 'hello'))
```

Results:

```python3
>>> video
Video(url='hello')
>>> video2
Video(url='hello')
>>> c
C(url=<__main__.FirstSuccessfulOf object at 0x7fab81073ca0>)
```

**Comment by nuno-andre** [comment]
Hi!

If I've understood the case correctly, maybe a factory could be a smoother approach than a builder. E.g. something such as:

```python
from collections import defaultdict, abc
from pydantic import BaseModel

class Model(BaseModel):

    @classmethod
    def take_first(cls, **kwargs) -> 'Model':
        errors = defaultdict(list)
        values = dict()

        for k, cand in kwargs.items():
            validate = cls.__fields__[k].validate

            if not isinstance(cand, abc.Iterable):
                cand = [cand]
            elif isinstance(cand, str):
                cand = [cand]

            for c in cand:
                r = c() if callable(c) else c
                v, err = validate(r, {}, loc=k)
                if err:
                    errors[k].append(err)
                else:
                    values[k] = v
                    break

        print(errors)
        return cls(**values)


class Video(Model):
    url: str


v = dict(url=[None, lambda: None, lambda: 'hello'])
video = Video.take_first(**v)
video2 = Video.take_first(url='hello')
```

Cheers!

**Comment by eddsalkield** [comment, marked as answer]
Thanks @nuno-andre for taking the time to look into this with me!  I've been reflecting on this recently, and I have some further thoughts.  Please don't take this as a rejection of your ideas, but rather an evolution of what I've been thinking based upon your answer.  I'll mark this issue as resolved, because now that I've crystallised an approach, it's probably best discussed over on the issue tracker.

Your idea of using a `classmethod` is a pretty elegant one; as it stands however, the logic of `Model.take_first` is built into the model, meaning that we'd have to create two different models if we wanted different evaluation strategies for the same type.  To resolve this, we could consider a more general method, `Model.generate`, which evaluates arbitrary generator functions for each of its arguments.  Each generator function would receive the relevant validator as an argument, and would be expected to either return a value that validates correctly, or some errors explaining why validation failed.

The other thing I'd like to be able to handle, which the current approach doesn't achieve, is lazily evaluated recursion: that generator functions (such as `FirstSuccessfulOf`) should be nestable within each other such that unnecessary generator functions are left unevaluated.  Therefore, you could define generators for nested data structures such as iterables and sub-models, only evaluating the _required_ generator functions in the process.

For example, in the current approach, if we had models like so:

```python3
class A(BaseModel):
    i: int

class B(BaseModel):
    a: A
```

Then we get into a bit of a mess if we try to define a nested instance of B:

```python3
b = B.generate(a = FirstSuccessfulOf(
                A.generate(i = FirstSuccessfulOf(None, None)),
                A.generate(i = 1)
        ))
```

In this example, I intend to express that in generating attribute `a` of `b`, two possible strategies can be tried.  The first, `A.generate(i = FirstSuccessfulOf(None, None))`, should fail - because setting `i` to None is invalid.  Then the second strategy, `A.generate(i = 1)`, should be tried, and succeed.  We want to ensure that the second strategy is only attempted if the first one fails.

However, since Python evaluates expressions eagerly, the innermost expression `A.generate(i=FirstSuccessfulOf(None, None))` is evaluated first.  This fails, and so the whole expression fails.  The second strategy is never even given a chance!

To solve this, we need the inner representation of `A` to remain unresolved until the appropriate moment within the `FirstSuccessfulOf` generation function.  A natural representation of an unresolved model instance would be a dict (after all, the data that creates the model instance is a splattered dict in the [`__init__` function of `BaseModel`](https://github.com/samuelcolvin/pydantic/blob/0bfc9302749284d48dc407cf78f75b9bb8828dc7/pydantic/main.py#L407)).  This dict would map attribute names to data structures potentially containing other generation functions.  Then we could write something like like:

```python3
b = B.generate({'a': FirstSuccessfulOf(
                {'i': FirstSuccessfulOf(None, None)},
                {'i': 1}
        )})
```

So `B.generate` is now responsible for recursively descending through the dict and creating the model, passing the appropriate validators into any generator functions it finds (such as `FirstSuccessfulOf`), before constructing the model based upon the results.  This `generate` function is starting to share a remarkable amount of behaviour with [`validate_model`](https://github.com/samuelcolvin/pydantic/blob/0bfc9302749284d48dc407cf78f75b9bb8828dc7/pydantic/main.py#L1010), which does essentially an identical job but without the logic to evaluate generator functions.  This similarity only gets more uncanny when we have to consider consolidating and flattening the errors that get returned from the generator functions.

Now I think that, with a relatively small modification, we can augment the existing behaviour of pydantic for this purpose.  Consider `validate_model`: for every `(key, value)` pair within the dict of data that it receives, it calls `ModelField.validate` on `value`, which recursively evaluates it, handling the errors.  Let's create a new type that represents generator functions, `PydanticGeneration` (for want of a better name).  `FirstSuccessfulOf` is implemented as an instance of `PydanticGeneration`.  Then we modify `ModelField.validate` such that if its argument is of type `PydanticGeneration`, then this function is first evaluated before continuing with the validation process.

We get recursion for free, because the `PydanticGeneration` generator function is passed the validator that calls it.  It can run this validator to choose which value to return at the end.  Notably, this validator may in itself contain a `PydanticGeneration` generator function.

The result is that we would be able to write something like the following:

```python3
b = B(a = FirstSuccessfulOf(
                {'i': FirstSuccessfulOf(None, None)},
                {'i': 1}
        ))
```

This evaluates in the order that we want, and also handles errors for us in a way that's pretty consistent with the rest of pydantic.  If you read this far, thank you! :sweat_smile:

I think overall what I've converged on here constitutes a pretty good approach for dynamically generating model attributes, but I'm open to other suggestions.  As I mentioned before, I think a good next step is opening an issue on the issue tracker and seeing whether the pydantic team would be open to a PR implementing behaviour similar to this.
